File: java/src/main/java/toolgood/algorithm/internals/functions/csharp/Function_REGEXREPLACE.java
package toolgood.algorithm.internals.functions.csharp;

import java.lang.StringBuilder;
import java.util.List;
import java.util.function.BiFunction;
import java.util.regex.Pattern;

import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.Function_3;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.NoneEngine;

public final class Function_REGEXREPLACE extends Function_3 {
    public Function_REGEXREPLACE(FunctionBase[] funcs) {
        super(funcs);
    }

    @Override
    public String Name() {
        return "RegexReplace";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        Operand args1 = GetText_1(engine, tempParameter);
        if (args1.IsError() || args1.IsNone()) {
            return args1;
        }

        Operand args2 = GetText_2(engine, tempParameter);
        if (args2.IsError() || args2.IsNone()) {
            return args2;
        }

        Operand args3 = GetText_3(engine, tempParameter);
        if (args3.IsError() || args3.IsNone()) {
            return args3;
        }

        try {
            String result = Pattern.compile(args2.TextValue()).matcher(args1.TextValue()).replaceAll(args3.TextValue());
            return Operand.Create(result);
        } catch (Exception e) {
            return ParameterError(2);
        }
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.TEXT;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result,
            OperandType operandType, String op, String val) {
        func1.GetParameterTypes(noneEngine, result, OperandType.TEXT, null, null);
        func2.GetParameterTypes(noneEngine, result, OperandType.TEXT, null, null);
        func3.GetParameterTypes(noneEngine, result, OperandType.TEXT, null, null);
    }
}

File: csharp/ToolGood.Algorithm/Internals/Functions/Csharp/Function_REGEXREPLACE.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using ToolGood.Algorithm.Enums;

namespace ToolGood.Algorithm.Internals.Functions.Csharp
{
	internal sealed class Function_REGEXREPLACE : Function_3
	{
		public Function_REGEXREPLACE(FunctionBase[] funcs) : base(funcs)
		{
		}

		public override string Name => "RegexReplace";

		public override Operand Evaluate(AlgorithmEngine engine, Func<AlgorithmEngine, string, Operand> tempParameter)
		{
			var args1 = GetText_1(engine, tempParameter);
			if(args1.IsErrorOrNone) { return args1; }

			var args2 = GetText_2(engine, tempParameter);
			if(args2.IsErrorOrNone) { return args2; }

			var args3 = GetText_3(engine, tempParameter);
			if(args3.IsErrorOrNone) { return args3; }

			try {
				var b = Regex.Replace(args1.TextValue, args2.TextValue, args3.TextValue);
				return Operand.Create(b);
			} catch (ArgumentException) {
				return ParameterError(2);
			}
		}
		public override OperandType GetResultType()
		{
			return OperandType.TEXT;
		}

		internal override void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, string op = null, string val = null)
		{
			func1.GetParameterTypes(noneEngine, result, OperandType.TEXT);
			func2.GetParameterTypes(noneEngine, result, OperandType.TEXT);
			func3.GetParameterTypes(noneEngine, result, OperandType.TEXT);
		}
	}

}
